web/build.sh

```diff
@@ -177,70 +230,204 @@ minify ( ) {
+      mkdir -p $CONFIG_OUT_PATH/src/$SOURCE_FOLDER
+      cp -Rf  $SOURCE/$SOURCE_FOLDER/    $CONFIG_OUT_PATH/src/$SOURCE_FOLDER
+    done
 
     echo
-    echo KeymanWeb resources saved under $1
-    echo
+  done
+}
+
+# Compiles compiled scripts from the first folder specified to the second folder specified.
+#
+# ### Parameters
+#
+# * 1: `src`      the folder containing scripts to be copied
+# * 2: `dst`      the destination folder for the copy operation
+# * 3: `scripts`  an array of filenames for expected compiled scripts
+#
+# ### Example
+#
+# ```bash
+#   compile_and_minify build/app/web/obj build/app/web/debug keymanweb.js
+# ```
+copy_outputs ( ) {
+  local src=$1
+  local dst=$2
+
+  shift
+  shift
+
+  local BASE_SCRIPTS=("$@")
+
+  mkdir -p "$dst"
+
+  for SCRIPTJS in "${BASE_SCRIPTS[@]}";
+  do
+    cp -Rf $src/$SCRIPTJS             $dst/
+    cp -Rf $src/$SCRIPTJS.map         $dst/
+  done
+}
+
+# Compiles all build products corresponding to the specified target.
+#
+# ### Parameters
+#
+# * 1: `product`    the product's source path under src/
+#
+# ### Example
+#
+# ```bash
+#   compile app/embed
+# ```
+compile ( ) {
+  if [ $# -lt 1 ]; then
+    fail "Scripting error: insufficient argument count!"
+  fi
+
+  local COMPILE_TARGET=$1
+  local COMPILED_INTERMEDIATE_PATH=$(output_path $COMPILE_TARGET $INTERMEDIATE)
+
+  shift
+
+  $compilecmd -b src/$COMPILE_TARGET -v
+
+  echo $COMPILE_TARGET TypeScript compiled under $COMPILED_INTERMEDIATE_PATH
+}
+
+# Finalizes all build products corresponding to the specified target.
+# This should be called after `compile` for all `app/` targets.
+#
+# ### Parameters
+#
+# * 1: `product`    the product's source path under src/
+# * 2: `outputs`    an array of expected output script files for the build
+#
+# ### Example
+#
+# ```bash
+#   compile app/embed
+#   finalize app/embed keyman.js
+# ```
+finalize ( ) {
+  if [ $# -lt 2 ]; then
+    fail "Scripting error: insufficient argument count!"
+  fi
+
+  local COMPILE_TARGET=$1
+  local COMPILED_INTERMEDIATE_PATH=$(output_path $COMPILE_TARGET $INTERMEDIATE)
+  local DEBUG_OUT_PATH=$(output_path $COMPILE_TARGET $DEBUG)
+  local RELEASE_OUT_PATH=$(output_path $COMPILE_TARGET $RELEASE)
+
+  shift
+
+  local OUTPUT_SCRIPTS=("$@")
+
+  # START:  debug output
+
+  mkdir -p "$DEBUG_OUT_PATH"
+  copy_outputs $COMPILED_INTERMEDIATE_PATH $DEBUG_OUT_PATH "${OUTPUT_SCRIPTS[@]}"
+
+  echo Compiled $COMPILE_TARGET debug version saved under $DEBUG_OUT_PATH: ${OUTPUT_SCRIPTS[*]}
+
+  # START:  release output
+  if ! builder_has_option --skip-minify; then
+    for SCRIPT in "${OUTPUT_SCRIPTS[@]}";
+    do
+      minify $COMPILED_INTERMEDIATE_PATH/$SCRIPT $RELEASE_OUT_PATH/$SCRIPT SIMPLE_OPTIMIZATIONS
+    done
+
+    echo Compiled $COMPILE_TARGET release version saved under $RELEASE_OUT_PATH: ${OUTPUT_SCRIPTS[*]}
+  else
+    # The prior 'release' is now outdated:  delete it.
+    rm -rf $RELEASE_OUT_PATH
```

Comment:
```suggestion
    rm -rf "$RELEASE_OUT_PATH"
```

Super important ... a space in the path here would cause havoc